commerce/domain/tax/cart/calculator/classes/TaxCartCalculatorSample.cls

```diff
@@ -1,349 +1,249 @@
-// CartExtension.TaxCartCalculator class.
+// This class facilitates tax calculation either by invoking an external service or by implementing the tax calculation logic internally.
+// The resulting tax values and adjustments are then stored in a Cart Data Transfer Object (DTO), references to both code implementations provided
+// For a tax calculator extension to be processed by the checkout flow, you must implement the CartExtension.TaxCartCalculator class.
+
 public class TaxCartCalculatorSample extends CartExtension.TaxCartCalculator {
 
-// You MUST change this to be your service or you must launch your own Third Party Service
-// and add the host in Setup | Security | Remote site settings.
-  private static String externalTaxHost = 'https://example.com';
-
-  // You MUST change the useExternalService to True if you want to use the Third Party Service.
-  private static Boolean useExternalService = false;
-  public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
-    try {
-      CartExtension.Cart cart = request.getCart();
-
-      // Clean up CVO based on tax. When new tax calculator request comes, we need to clean up
-      // previous CVOs as they have been previously handled by the Cart Calculate API.
-      CartExtension.CartValidationOutputList cartValidationOutputCollection = cart.getCartValidationOutputs();
-      Iterator<CartExtension.CartValidationOutput> cartValidationOutputCollectionIterator = cartValidationOutputCollection.iterator();
-      while (cartValidationOutputCollectionIterator.hasNext()) {
-        CartExtension.CartValidationOutput cvo = cartValidationOutputCollectionIterator.next();
-        if (cvo.getType() == CartExtension.CartValidationOutputTypeEnum.TAXES) {
-          cartValidationOutputCollection.remove(cvo);
-        }
-      }
+	// You MUST change this to be your service or you must launch your own Third Party Service and add the host in Setup | Security | Remote site settings.
+	private static String externalTaxHost = 'https://example.com';
+
+	// You MUST change the useExternalService to True if you want to use the Third Party Service.
+	private static Boolean useExternalService = false;
+
+	public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
+		CartExtension.Cart cart = request.getCart();
+		CartExtension.CartValidationOutputList cartValidationOutputCollection = cart.getCartValidationOutputs();
+		try{
+			Integer cartItemIdSeq = 0;
+			if (!isValidCart(cart)){
+                CartExtension.CartValidationOutput cvo = new CartExtension.CartValidationOutput(
+                CartExtension.CartValidationOutputTypeEnum.TAXES,
+                CartExtension.CartValidationOutputLevelEnum.ERROR);
+                cvo.setMessage('There was a problem with delivery address in request.');
+                cartValidationOutputCollection.add(cvo);
```

Comment:
This code is not idempotent. If you execute the same code on the same cart twice - 2 CVOs are created instead of one. (Please add a unit test for that)

Reply:
At starting of code for every request whether it is for same or different cart, removing all cvos related to taxes to ensure cvos list is empty and so that it won't create duplicate entries